
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Exception - C++ - W3cubDocs</title>
  
  <meta name="description" content=" This header is part of the error handling library. ">
  <meta name="keywords" content="standard, library, header, exception, c++, cpp">
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/cpp/header/exception.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-e4ebd3a2a5652ff55173659804c4390a004917f3bdd17b5bb3ba78ea5c9c46fe181cadaac34517ccd815f5bdc982bbfe67179d6f4ac2f084ef2265e2a3dc8dc5.css" integrity="sha512-5OvToqVlL/VRc2WYBMQ5CgBJF/O90Xtbs7p46lycRv4YHK2qw0UXzNgV9b3Jgrv+Zxedb0rC8ITvImXio9yNxQ==" crossorigin="anonymous">
  <script type="text/javascript" integrity="sha512-EpkDeu98lN/jPKijllzVWdRg/dUSSMCaldYZNFz6bcNoBvpWRNz0HSTRQJ3ENmQc5Cuj1zDW1vHd7b0DzpOgyA==" crossorigin="anonymous" src="/assets/application-1299037aef7c94dfe33ca8a3965cd559d460fdd51248c09a95d619345cfa6dc36806fa5644dcf41d24d1409dc436641ce42ba3d730d6d6f1ddedbd03ce93a0c8.js"></script>
  <script src="/json/cpp.js"></script>
  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-71174418-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>
  <script data-ad-client="ca-pub-2572770204602497" async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body>
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">
  
  <form class="_search">
    <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
    <a class="_search-clear"></a>
    <div class="_search-tag"></div>
  </form>
  
  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/cpp/" class="_nav-link" title="" style="margin-left:0;">C++</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="link"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _c">
				
				
<h1 id="firstHeading" class="firstHeading">Standard library header &lt;exception&gt;</h1>            <p>This header is part of the <a href="../error" title="cpp/error">error handling</a> library.</p>
<table class="t-dsc-begin"> <tr> <td colspan="2"> <h3 id="Classes">  Classes</h3> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/exception" title="cpp/error/exception"> <span class="t-lines"><span>exception</span></span></a></div> </td> <td> base class for exceptions thrown by the standard library components <br> <span class="t-mark">(class)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/nested_exception" title="cpp/error/nested exception"> <span class="t-lines"><span>nested_exception</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span></span></span></div> </td> <td> a mixin type to capture and store current exceptions <br> <span class="t-mark">(class)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/bad_exception" title="cpp/error/bad exception"> <span class="t-lines"><span>bad_exception</span></span></a></div> </td> <td> exception thrown when <code><a href="../error/current_exception" title="cpp/error/current exception">std::current_exception</a></code> fails to copy the exception object <br> <span class="t-mark">(class)</span> </td>
</tr> <tr> <td colspan="2"> <h3 id="Typedefs">  Typedefs</h3> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/exception/unexpected_handler" title="cpp/error/unexpected handler"> <span class="t-lines"><span>unexpected_handler</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-until-cxx17">(removed in C++17)</span></span></span></div> </td> <td> the type of the function called by <code><a href="../error/exception/unexpected" title="cpp/error/unexpected">std::unexpected</a></code> <br> <span class="t-mark">(typedef)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/terminate_handler" title="cpp/error/terminate handler"> <span class="t-lines"><span>terminate_handler</span></span></a></div> </td> <td> the type of the function called by <code><a href="../error/terminate" title="cpp/error/terminate">std::terminate</a></code> <br> <span class="t-mark">(typedef)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/exception_ptr" title="cpp/error/exception ptr"> <span class="t-lines"><span>exception_ptr</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span></span></span></div> </td> <td> shared pointer type for handling exception objects <br> <span class="t-mark">(typedef)</span> </td>
</tr> <tr> <td colspan="2"> <h3 id="Functions">  Functions</h3> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/exception/unexpected" title="cpp/error/unexpected"> <span class="t-lines"><span>unexpected</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-until-cxx17">(removed in C++17)</span></span></span></div> </td> <td> function called when dynamic exception specification is violated <br> <span class="t-mark">(function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/exception/uncaught_exception" title="cpp/error/uncaught exception"> <span class="t-lines"><span>uncaught_exception</span><span>uncaught_exceptions</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark">(deprecated in c++17)</span><span class="t-mark-rev t-until-cxx20">(removed in C++20)</span></span><span><span class="t-mark-rev t-since-cxx17">(C++17)</span></span></span></div> </td> <td> checks if exception handling is currently in progress <br> <span class="t-mark">(function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/make_exception_ptr" title="cpp/error/make exception ptr"> <span class="t-lines"><span>make_exception_ptr</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span></span></span></div> </td> <td> creates an <code><a href="../error/exception_ptr" title="cpp/error/exception ptr">std::exception_ptr</a></code> from an exception object <br> <span class="t-mark">(function template)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/current_exception" title="cpp/error/current exception"> <span class="t-lines"><span>current_exception</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span></span></span></div> </td> <td> captures the current exception in a <code><a href="../error/exception_ptr" title="cpp/error/exception ptr">std::exception_ptr</a></code> <br> <span class="t-mark">(function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/rethrow_exception" title="cpp/error/rethrow exception"> <span class="t-lines"><span>rethrow_exception</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span></span></span></div> </td> <td> throws the exception from an <code><a href="../error/exception_ptr" title="cpp/error/exception ptr">std::exception_ptr</a></code> <br> <span class="t-mark">(function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/throw_with_nested" title="cpp/error/throw with nested"> <span class="t-lines"><span>throw_with_nested</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span></span></span></div> </td> <td> throws its argument with <code><a href="../error/nested_exception" title="cpp/error/nested exception">std::nested_exception</a></code> mixed in <br> <span class="t-mark">(function template)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/rethrow_if_nested" title="cpp/error/rethrow if nested"> <span class="t-lines"><span>rethrow_if_nested</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span></span></span></div> </td> <td> throws the exception from a <code><a href="../error/nested_exception" title="cpp/error/nested exception">std::nested_exception</a></code> <br> <span class="t-mark">(function template)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/terminate" title="cpp/error/terminate"> <span class="t-lines"><span>terminate</span></span></a></div> </td> <td> function called when exception handling fails <br> <span class="t-mark">(function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/get_terminate" title="cpp/error/get terminate"> <span class="t-lines"><span>get_terminate</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span></span></span></div> </td> <td> obtains the current terminate_handler <br> <span class="t-mark">(function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/set_terminate" title="cpp/error/set terminate"> <span class="t-lines"><span>set_terminate</span></span></a></div> </td> <td> changes the function to be called by <code><a href="../error/terminate" title="cpp/error/terminate">std::terminate</a></code> <br> <span class="t-mark">(function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/exception/get_unexpected" title="cpp/error/get unexpected"> <span class="t-lines"><span>get_unexpected</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx11">(C++11)</span><span class="t-mark-rev t-until-cxx17">(removed in C++17)</span></span></span></div> </td> <td> obtains the current unexpected_handler <br> <span class="t-mark">(function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><a href="../error/exception/set_unexpected" title="cpp/error/set unexpected"> <span class="t-lines"><span>set_unexpected</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-until-cxx17">(removed in C++17)</span></span></span></div> </td> <td> changes the function to be called by <code><a href="../error/exception/unexpected" title="cpp/error/unexpected">std::unexpected</a></code> <br> <span class="t-mark">(function)</span> </td>
</tr> </table> <h3 id="Synopsis">Synopsis</h3> <div class="cpp source-cpp"><pre data-language="cpp">namespace std {
    class exception;
    class bad_exception;
    class nested_exception;
 
    typedef void (*unexpected_handler)();
    unexpected_handler get_unexpected() noexcept;
    unexpected_handler set_unexpected(unexpected_handler f) noexcept;
    [[noreturn]] void unexpected();
 
    typedef void (*terminate_handler)();
    terminate_handler get_terminate() noexcept;
    terminate_handler set_terminate(terminate_handler f) noexcept;
    [[noreturn]] void terminate() noexcept;
 
    bool uncaught_exception() noexcept;
 
    typedef /*unspecified*/ exception_ptr;
 
    exception_ptr current_exception() noexcept;
    [[noreturn]] void rethrow_exception(exception_ptr p);
    template&lt;class E&gt; exception_ptr make_exception_ptr(E e) noexcept;
 
    [[noreturn]] template &lt;class T&gt; void throw_with_nested(T&amp;&amp; t);
    template &lt;class E&gt; void rethrow_if_nested(const E&amp; e);
}</pre></div> <h4 id="Class_std::exception">Class <code><a href="../error/exception" title="cpp/error/exception">std::exception</a></code>
</h4> <div class="cpp source-cpp"><pre data-language="cpp">class exception {
public:
    exception() noexcept;
    exception(const exception&amp;) noexcept;
    exception&amp; operator=(const exception&amp;) noexcept;
    virtual ~exception();
    virtual const char* what() const noexcept;
};</pre></div> <h4 id="Class_std::bad_exception">Class <code><a href="../error/bad_exception" title="cpp/error/bad exception">std::bad_exception</a></code>
</h4> <div class="cpp source-cpp"><pre data-language="cpp">class bad_exception : public exception {
public:
    bad_exception() noexcept;
    bad_exception(const bad_exception&amp;) noexcept;
    bad_exception&amp; operator=(const bad_exception&amp;) noexcept;
    virtual const char* what() const noexcept;
};</pre></div> <h4 id="Class_std::nested_exception">Class <code><a href="../error/nested_exception" title="cpp/error/nested exception">std::nested_exception</a></code>
</h4> <div class="cpp source-cpp"><pre data-language="cpp">class nested_exception {
public:
    nested_exception() noexcept;
    nested_exception(const nested_exception&amp;) noexcept = default;
    nested_exception&amp; operator=(const nested_exception&amp;) noexcept = default;
    virtual ~nested_exception() = default;
 
    // access functions
    [[noreturn]] void rethrow_nested() const;
    exception_ptr nested_ptr() const noexcept;
};</pre></div> <h3 id="See_Also">See Also</h3> <ul><li>
<a href="../error" title="cpp/error">Error handling</a> </li></ul>          <div class="_attribution">
  <p class="_attribution-p">
    © cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="http://en.cppreference.com/w/cpp/header/exception" class="_attribution-link">http://en.cppreference.com/w/cpp/header/exception</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
